### ISSUE TITLE
Getting UV coordinates of representations

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#2021]
Right now, it seems there is no UV support such as for IfcIndexedTriangleTextureMap in the geom iterator / create_shape. This means that we'd either bypass the iterator or kinda reverse engineer UV coords afterwards (which probably wouldn't be super successful - one potentially welded vertex can correlate to multiple UV coordinates, equivalent to the number of polygons it is part of).

Ideally, the support would need to be built into C++. But I can also understand this would be tricky.

Here are some C++ locations I think would make sense to start looking, but I have no idea if I know enough to succeed here.

 - This triangulation function should capture uvs similar to how it captures material_ids https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom/IfcGeomRepresentation.h#L139
 - Maybe a new UVs parameter here? https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom/IfcRepresentationShapeItem.h#L29
 - Make sure UVs propagate through mapped items by bubbling the uvs here: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom/IfcGeomShapes.cpp#L840
 - `create_brep_for_representation_and_product` to be updated to include UVs. Is this the right spot to actually grab the UV coordinates to begin with? https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom/IfcGeomFunctions.cpp#L2157

**Comment by aothms** [comment]
Note that this also applies to indexedcolourmap, so whatever we do we should stick to the same (general) mechanism of overlaying additional face information channels (both colour and uv are defined in the context of a face, not vertex in IFC).

The same would apply to IfcReparametrisedCompositeCurveSegment because I think trying to shoehorn that into the geometric primitives will cause a lot of headaches, better as a separate uv channel I reckon. Although I'm not entirely sure, thinking about IfcPointOnSurface kind of things.

Ideally shapes would also need to be able to undergo boolean operations and have the uv coords and colour properly propagated/interpolated.

Those concern would make me think that we should try and store this somehow using opencascade's mechanism. And that would be https://dev.opencascade.org/doc/refman/html/class_t_d_f___attribute.html probably as a RealArray This is something I have postponed looking into for 10 years, and I'm not eager to dive into it now.

Maybe it's quickest if we try and asses the TDF_Attribute thing outside of IfcOpenShell, create some quick prototypes and see if it covers our use cases and survives extrusion, boolean and meshing. That way it'd be easier maybe to get feedback and quickly iterate until we have validated a proper way forward.

**Comment by Moult** [comment]
@aothms I can help provide sample files, now that we can export UV textures out from Blender. I'd also like to help where possible, but you know best which portions of C++ need doing, and which are easier and which aren't.

Why is IfcReparametrisedCompositeCurveSegment involved? Can't UVs simply be used for triangulated elements and ignored (or set to 0,0) for everything else as a starting point?

**Comment by aothms** [comment]
> Why is IfcReparametrisedCompositeCurveSegment involved? Can't UVs simply be used for triangulated elements and ignored (or set to 0,0) for everything else as a starting point?

For the time being sure, but conceptually we need a solution to handle all cases.

Another case where this is relevant btw is that we still can't support a StyledByItem on a IfcFace part of a shell. So if we use occt attributes that'd be another thing we should look into.

**Comment by Moult** [comment]
For my own notes, here's the most efficient way of inserting UVs into Blender with foreach_set: https://blenderartists.org/t/importing-uv-coordinates/595872/5 also [good reading](https://b3d.interplanety.org/en/working-with-uv-maps-through-the-blender-api/) and helpful to visualise UVs and how they correlate to loops.

I haven't yet grokked the foreach_set trick for this so I can't figure out yet what the ideal flattened list end output would be (at least for Blender, but I assume it'd be similar for other apps). I'll try re-reading it with a fresh set of eyes tomorrow.

Is there anything I can help with this?

**Comment by aothms** [comment]
Your knowledge on TDF_Attribute is as good as mine so if you feel like giving this a jump start feel free to have al look on how it can be used to associate the uvs with the shape. Feel free to disregard boolean and extrusion for now. Main questions:

* what does the code actually look like, in terms of writing (in IfcGeomShapes::convert(IfcTriangulatedFaceset)) and reading (IfcGeomRepresentation.h) 
* how to conceptually store the data, from what I read the TDF naming thing is a tree, so you can indeed store data associated to a vertex as it is reachable from a face, even if the vertex is shared by multiple faces.

**Comment by Moult** [comment]
I'll try my best. It looks like voodoo right now to me. Is there a developer guide somewhere to where things live in the c++ codebase? 
I've been doing a lot of grepping and guessing but I'm still missing an overall understanding of structure of the logic flow when create shape is called from beginning until end. Perhaps I can try to write one whilst trying to solve this - learning c++, writing docs (aimed at Python programmers perhaps who don't know c++)?

**Comment by aothms** [comment]
For understanding the logic flow I'd set a breakpoint in `IfcGeomShapes::convert(IfcTriangulatedFaceset))` and look at the call stack frames. There's quite a bit of indirection in selecting the schema and bubbling up to the correct type conversion function. But the call stack shows it reasonable well.

But like I said, maybe with a prototype outside of IfcOpenShell we can iterate fastest and invite more people for feedback.

**Comment by Gorgious56** [comment]
> For my own notes, here's the most efficient way of inserting UVs into Blender with foreach_set: https://blenderartists.org/t/importing-uv-coordinates/595872/5 also [good reading](https://b3d.interplanety.org/en/working-with-uv-maps-through-the-blender-api/) and helpful to visualise UVs and how they correlate to loops.
> 
> I haven't yet grokked the foreach_set trick for this so I can't figure out yet what the ideal flattened list end output would be (at least for Blender, but I assume it'd be similar for other apps). I'll try re-reading it with a fresh set of eyes tomorrow.

I know you already know most if not all of this stuff but I'll expand on it in case it's somehow helpful :

As I understand it in Blender at least the UVs are defined much like the vertex colors. Each vertex actually holds Ni uv coordinates pairs where Ni is the number of polygons it is a part of. You can estimate the number of vertex coordinates pairs by looking at the face angles, because each face angle is associated to what's called a loop in Blender :

![image](https://user-images.githubusercontent.com/25156105/152785312-9c930a3b-e1cb-4adb-827b-d9deaf2dd350.png)

When trivially associating UVs to a mesh (basic unwrapping function), all the uv pairs of each vertex contain the same coordinates. A vertex connected to 3 faces on the mesh will appear only once on the UV map. However when texturing an object you often want to add seams, for instance when two polygons meet at an edge and you want the texture to be displayed differently, or to avoid distortion when unwrapping a cylinder. Then a vertex can appear more than one time in the UV map, where each point is represented by its associated UV loop coordinates.

Here's a visualization of the central vertex having 3 associated vertex colors (which may aswell be uvw coordinates) associated to each one of its loops : Green for top face, red for lower left face, blue for lower right face.

![image](https://user-images.githubusercontent.com/25156105/152786123-9d949fec-5186-4b19-b516-4b5d427a582b.png)

A [mesh in Blender](https://docs.blender.org/api/current/bpy.types.Mesh.html#mesh-id) holds data for several entities : [Vertices](https://docs.blender.org/api/current/bpy.types.MeshVertex.html#bpy.types.MeshVertex) (list of x, y, z coordinates), [Edges](https://docs.blender.org/api/current/bpy.types.Mesh.html#bpy.types.Mesh.edges) (list of pairs of vertex indices), [Polygons](https://docs.blender.org/api/current/bpy.types.MeshPolygon.html#bpy.types.MeshPolygon) (list of list of vertex indices, AND list of loop indices), and [Loops](https://docs.blender.org/api/current/bpy.types.MeshLoop.html#bpy.types.MeshLoop) (list of edge index + vertex index). 

So a polygon contains a list of loops which can be uniquely associated to a given vertex and edge on the mesh parent data block. Then I assume that the ideal flattened list would be a list which would associate for each loop in `mesh.loops` the uv coordinates of the associated vertex and edge. Its size would be 2 (uv coordinates) * Number of Polygons * Number of vertices in each polygon :


```
import bpy

mesh = bpy.data.objects["Cube"].data

def uv_flatten(mesh):
    for polygon in mesh.polygons:
        for loop in polygon.loop_indices:
            yield from mesh.uv_layers.active.data[loop].uv

print(list(uv_flatten(mesh)))
```

And we get our very, very, very understandable list comprehension from that : 

`
print([channel for coords in [mesh.uv_layers.active.data[index].uv for indices in [polygon.loop_indices for polygon in mesh.polygons] for index in indices] for channel in coords])
`

(I'd rather use for loops, thanks :) )

**Comment by Moult** [comment]
Thanks for the writeup @Gorgious56 - it will be valuable especially to others new to the concept too. I've written code similar to yours for extracting UVs, but one thing I didn't work out yet (I was planning to revisit it with a fresh set of eyes) was the format required for _setting_  UVs using foreach_set (which presumably needs a flattened list of some sort). Is that flattened list the list comprehension you wrote at the end?

**Comment by Gorgious56** [comment]
@Moult  Yes, I think this is what the flattened list should look like, in fact here's a code snippet to test it :

```
import bpy
import numpy as np

mesh = bpy.data.objects["Cube"].data
uvs = np.empty(shape=len(mesh.uv_layers.active.data) * 2, dtype=np.float32)
mesh.uv_layers.active.data.foreach_get("uv", uvs)
print(uvs)
mesh.uv_layers.active.data.foreach_set("uv", uvs)  # Simple roundtrip

def uv_flatten(mesh):
    for polygon in mesh.polygons:
        for loop in polygon.loop_indices:
            yield from mesh.uv_layers.active.data[loop].uv
        
for i, uv in enumerate(uv_flatten(mesh)):
    uvs[i] = uv
print(uvs)
mesh.uv_layers.active.data.foreach_set("uv", uvs)  # Same UV map !
```

**Comment by wolfboyyang** [comment]
I recently made some effort to support IfcIndexedTriangleTextureMap & IfcIndexedPolygonalTextureMap. There're several issues 
1. should_clean_mesh option would merge triangles into quads
```python
should_clean_mesh: BoolProperty(name="Clean Meshes", default=True)
```
to check the original triangles: change should_clean_mesh default to False. (There should be one options to disable clean mesh when load ifc project, I hope this could appear in advanced option while loading.)

2. quads are triangulated when creating geometries for IfcPolygonalFaceSet
This part is writen in C++ part of ifcopenshell. I didn't figure out why this happens yet. `bpy.ops.mesh.tris_convert_to_quads()` would help to merge the triangles, with should_clean_mesh, but should merge before setting the uv.

3. vertex indices changed when creating the triangle face or polygonal face.
the vertex indices in Blender, ifcopenshell, ifc file have diferent orders. ifcopenshell does not create stable vertex order, which means by saving and loading ifc file, the face indices may change. So, I need to find a way to sort the indices and find the correct face and vertex to apply TextureMap. But as coordinate in CoordList may duplicate [e.g., Sample-Test-Files tessellation-with-image-texture.ifc](https://github.com/buildingSMART/Sample-Test-Files/blob/master/IFC%204.0/BuildingSMARTSpec/tessellation-with-image-texture.ifc), sort indices in may not solve the problem. (Luckily, the face list seems remaining the same as in the file for IfcTriangulatedFaceSet.)

Display indices may help to check this.
[Display indices in Editing Mode](https://blender.stackexchange.com/questions/158493/displaying-vertex-indices-in-blender-2-8-using-debug-mode)


You could try with this Patch file for blenderbim-230711 build.
[texture_support_blenderbim-230711.patch](https://github.com/IfcOpenShell/IfcOpenShell/files/12016005/texture_support_blenderbim-230711.patch)

Test files
[test-tessellation-with-image-texture.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/12016060/test-tessellation-with-image-texture.zip)

<img width="1504" alt="test_tessellation-with-image-texture_IfcIndexedPolygonalTextureMap_4x3" src="https://github.com/IfcOpenShell/IfcOpenShell/assets/6252922/52c61b3a-6d55-4afc-b4d1-118010bdda9d">

<img width="1512" alt="tessellation-with-image-texture" src="https://github.com/IfcOpenShell/IfcOpenShell/assets/6252922/e987193c-441f-4716-a889-99012193e874">

**Comment by Moult** [comment]
Huh, very cool! It seems this is the secret sauce:

```
+        # sort the verts to orginal order in ifc file
+        bm.verts.sort(key=lambda v: coords.index(v.co))
```


The reason I avoid doing this is because I was always afraid of how to deal with dynamic voids in meshes was was hoping IfcOpenShell could reparameterise the UV to the newly generated vertices around the voids. This means that there could be additional vertices that the IFC doesn't have, and therefore to solve it it would need to be done in the C++ level. But the more I think about it perhaps it isn't a practical concern.

So for point 1, completely agree. Clean mesh needs to be disabled. Point 2, IfcOpenShell tessellates the resulting OpenCASCADE BRep which gives us triangles. There is some progress done on a CGAL or hybrid kernel implementation which would allow us to preserve exactly the original faces of the IFC mesh, but your solution will be a very welcome stopgap until that happens.

Would you like to submit this as a PR or are you happy for me to test this further and integrate your code? (Of course I will credit you in commit logs)

**Comment by wolfboyyang** [comment]
@Moult Thanks for your explaination. You could go ahead to test this further and integrate the code. I don't mind. If you're busy with other features, I could also try to create PR (But as I'm travelling, I may not have much time until next week).
